File: search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def searchUsingBinarySearch(left, right):
            while(left<= right):
                middle = (left+right)//2
                if(nums[middle] == target):
                    return middle
                elif(nums[middle]> target):
                    right = middle-1
                else:
                    left = middle+1
            return -1
        
        def searchForPivot(left, right):
            middle = (left+right)//2    
            if(middle+1 < len(nums)):
                if(nums[middle] >= nums[middle+1]):
                    return middle
            elif(middle-1>=0):
                if(nums[middle-1] >= nums[middle]):
                    return middle-1
            if(right> left):
                val1 = searchForPivot(left, middle-1)
                val2 = searchForPivot(middle+1, right)
                
                if(val1 > -1):
                    return val1
                if(val2 > -1):
                    return val2
            return -1
        
        
        pivot = searchForPivot(0, len(nums)-1)
        if(pivot == -1):
            return searchUsingBinarySearch(0, len(nums)-1)
        if(nums[pivot] == target):
            return pivot
        
        val1 = searchUsingBinarySearch(0, pivot-1)
        val2 = searchUsingBinarySearch(pivot+1, len(nums)-1)
        return val1 if val1>-1 else val2    
        return -1
          
        

File: search-in-rotated-sorted-array/search-in-rotated-sorted-array.go
import ("fmt")

func findPeak(nums []int, left int, right int) int {
    size := len(nums)
    mid := (left+right)/2
    if left > right {
         return -1
    }
    if mid-1 >=0 && nums[mid] < nums[mid-1]  {
        return mid
    } else if mid+1 < size && nums[mid] > nums[mid+1] {
        return mid+1
    }
    leftSideIndex := findPeak(nums, left, mid-1)
    rightSideIndex := findPeak(nums, mid+1, right)

    if leftSideIndex > -1 {
        return leftSideIndex
    }
    if rightSideIndex > -1 {
         return rightSideIndex
    } 
    return -1
}
func seachInSortedArray(nums []int, left int, right int, target int) int {
    for left <= right {
        mid := (left + right)/2
        if nums[mid] == target {
            return mid
        } else if nums[mid] > target {
            right = mid-1
        } else {
            left = mid+1
        }
    }
    return -1
}


func search(nums []int, target int) int {
    size := len(nums)
    peakIndex := findPeak(nums, 0, size-1)
    fmt.Println(peakIndex)
    if peakIndex == -1 {
        case1 :=  seachInSortedArray(nums, 0, size-1, target)
        return case1
    } else {
        case1 :=  seachInSortedArray(nums, 0, peakIndex-1, target)
        case2 :=  seachInSortedArray(nums, peakIndex, size-1, target)
        if case1 > -1 {
        return case1
        } else if case2 > -1 {
            return case2
        } 
    }
    
    return -1
}